Refactor TS integration and convert Camera to TS

naftalibeder opened the pull request:
I've been wanting to do this for a while, because I think it will really speed up and simplify development work. This also changes the camera from a class to a functional component, and brings in modern memoization techniques.

It's a work in progress at the moment, but prop-based updates work perfectly. ~Just re-implementing the imperative methods.~ Those are done too.

### Checklist
- [X] Component
- [X] Types
- [x] Tests
- [x] Docs

naftalibeder commented:
@mfazekas I also wanted to check about documentation - the Camera props are typed and `/** documented */` in the file, but the generation script doesn't seem to be reading them. I haven't looked in much detail, but any ideas offhand?

mfazekas commented:
@naftalibeder we've only looking for  the `.js` files in doc generators. But they might need other changes as well.

https://github.com/rnmapbox/maps/blob/dda13db0c723c5f7e84341176cee77cd6897eb6c/scripts/autogenHelpers/DocJSONBuilder.js#L41-L46

naftalibeder commented:
@mfazekas This is working well, but there are few bigger blockers - namely tests and documentation.

Tests: The `Camera` tests are failing because the tests no longer apply to the component's implementation. This needs to get updated to match (it would have to test props/renders of the component, not the output of a function).

Docs: This is even worse! Some simple changes to include `ts|tsx` file in `DocJSONBuilder.js` made it clear that this would be a challenging road to go down. I also tried using `tsdoc`, but was having some bad luck with it failing to exclude directories properly.

I'm not sure how much time I can put into tests and documentation right now - am I correct in thinking that you wouldn't want to merge this until those are both complete?

naftalibeder commented:
I know [this last commit](https://github.com/rnmapbox/maps/pull/1872/commits/ea8a675c2c93f4fe345340f166fc3e0854bd354f) touches a number of important files, but I think it's a necessary change given that this is the first component to have 'real' type definitions (vs. the `.d.ts` definitions).

So we need a way to let the two kinds of type definitions exist in tandem. This seems to work. Let me know what you think.

mfazekas commented:
> I know [this last commit](https://github.com/rnmapbox/maps/pull/1872/commits/ea8a675c2c93f4fe345340f166fc3e0854bd354f) touches a number of important files, but I think it's a necessary change given that this is the first component to have 'real' type definitions (vs. the `.d.ts` definitions).
> 
> So we need a way to let the two kinds of type definitions exist in tandem. This seems to work. Let me know what you think. (It shouldn't break imports for anyone, as far as I can tell.)

Is my understanding is correct that this way we can write components with type definitions, and rest of the components types still coming from `definitions.ts`?! This sounds awesome!

mfazekas commented:
I've added a PR that demonstrates the changes needed to get doc generation work with tsx components.

https://github.com/TruckMap/maps/pull/158

naftalibeder commented:
@mfazekas I pulled in your changes manually, and added the rest of the prop documentation comments, and in general it's working quite well. I had to make some additional changes to format e.g. union types properly.

There are still some niceties we could add - e.g. formatting function and object type definitions, and `@example` blocks, using `<pre>` (I couldn't easily find where the backtick gets added around the displayed markdown type, to make that conditional.)

Based on [one change](https://github.com/rnmapbox/maps/pull/1872/commits/f5202d29d4d88ae0c5fe1231f93b08c532e94bf9#diff-72b8abdd8d19462ad4b41c4aace51f8ab8cbc8663b61f47238ac173d33a4aa52) to the template, a lot of other doc files would be changed. I changed it because I couldn't ignore the run-on sentence, and because not all components are `.js` now. Do you prefer I revert that file, or can I keep it in and update the rest of the docs?

mfazekas commented:
> @mfazekas I pulled in your changes manually, and added the rest of the prop documentation comments, and in general it's working quite well. I had to make some additional changes to format e.g. union types properly.
> 
> There are still some niceties we could add - e.g. formatting function and object type definitions, and `@example` blocks, using `\<pre\>` (I couldn't easily find where the backtick gets added around the displayed markdown type, to make that conditional.)
> 
> Based on [one change](https://github.com/rnmapbox/maps/pull/1872/commits/f5202d29d4d88ae0c5fe1231f93b08c532e94bf9#diff-72b8abdd8d19462ad4b41c4aace51f8ab8cbc8663b61f47238ac173d33a4aa52) to the template, a lot of other doc files would be changed. I changed it because I couldn't ignore the run-on sentence, and because not all components are `.js` now. Do you prefer I revert that file, or can I keep it in and update the rest of the docs?

Looks better. We still missing method docs, I'll look into if we can generate then. I think it's fine to update that comment in all the docs.

naftalibeder commented:
Added some initial (passing) tests ^

mfazekas commented:
Unfortunately I wasn't able to get doc generation fully working yet. 
The patch https://github.com/reactjs/react-docgen/pull/591, adds support for documenting methods in react-docgen. But it doesn't supports useMemo. Also doesn't gets methods from the typescript interface.

Probably the easiest workaround is adding some tags to method documentation and using reactNative handlers to extract that.

naftalibeder commented:
> adding some tags to method documentation

I'm not exactly sure what you mean - do you have an example that might work? Otherwise for the time being I could just document the imperative methods in the description.

mfazekas commented:
@naftalibeder I wasn't able to write handlers that extract directly from docs. 

I've added, which is main of react-docgen with https://github.com/reactjs/react-docgen/pull/592 and https://github.com/reactjs/react-docgen/pull/591

https://github.com/rnmapbox/react-docgen/tree/rnmapbox


Unforutantely I wasn't able to use my fork of `react-docgen` with yarn install as `prepare` scripts were not invoked and distributed folder was not created.

When I've moved methods docs to insider useImperativeHandler then it worked:
https://github.com/rnmapbox/maps/commit/d7efb8b96d8cdb378dbcd6f1b8fb4cf6c9220c08

naftalibeder commented:
@mfazekas Thanks for the example commit! It's working great, see my last 2 commits.

Thoughts? What do you think is left before this can get merged?

mfazekas commented on package.json:
```diff
@@ -6,7 +6,7 @@
   },
   "author": "React Native Mapbox Team",
-  "main": "./javascript/index.js",
+  "main": "index.ts",
```
Won't it cause issue for non ts app consuming it?!

Also don't we need a invoke typescript compile in `prepare` script?!

naftalibeder replied:
Absolutely, my bad - working on it.

naftalibeder replied:
Just FYI, this has been taking longer than I expected. Still in progress.

mfazekas replied:
@naftalibeder yep this has much more implications than I thought. But once we go over the first one changing the rest will be easy.

mfazekas commented on docs/Camera.md:
```diff
@@ -1,161 +1,122 @@
+<!-- Autogenerated from Camera. Do not modify. -->
 ## <MapboxGL.Camera />
-### 
+### Controls the perspective from which the user sees the map.<br/><br/>To use imperative methods, pass in a ref object:<br/><br/><pre>const camera = useRef<CameraRef>(null);<br/><br/>useEffect(() => {<br/>  camera.current?.setCamera({<br/>    centerCoordinate: [lon, lat],<br/>  });<br/>}, []);<br/><br/>return (<br/>  \<Camera ref={camera} /><br/>);</pre>
```
Previously they just used `useRef<Camera>`. Are we supporting that too?!

mfazekas commented on javascript/components/Camera.tsx:
```diff
@@ -0,0 +1,545 @@
+  // @ts-ignore
+  const camera: React.RefObject<RCTMGLCamera> = useRef(null);
+
+  const nativeAnimationMode = useCallback(
```
It should mirror the current version of `Camera.js` see default branch and handing of `none`.
https://github.com/rnmapbox/maps/blob/1049ab76957994ccfc837d6d53a8b1dc800c0f3f/javascript/components/Camera.js#L610-L624.



Btw we could just move this function out of `Camera` then we would not need `useCallback`.

naftalibeder commented:
I think I've got everything working with type declarations and compilation! Going to look over your other comments and then I think we'll be ready to review and test further.

naftalibeder commented:
@mfazekas Here's the current status:

**What's going well**

- The main project builds all TS and JS into `./lib` as part of the `generate` script. The output is a collection of `.js` and `.d.ts` file pairs. All file paths should be the same as before.
- The main project's `package.json` references `./lib` as the module start location.
- The `example` project is set up to reference the parent package more visibly, with the file path `..` in `example/package.json`.
- Types (not components; see below) are exported and can be correctly referenced.

**What's not going well**
- The default export (`MapboxGL`) Is not currently typed, so all of its properties are `any`. I haven't fixed this yet because... (see next).
- The component type declarations are not working quite right. Components that start as TS (`Camera.tsx`) have correctly typed props, but ones that start as JS (e.g. `MapView.js`) give the error `JSX element class does not support attributes because it does not have a 'props' property.`. I'm avoiding errors right now by referencing `MapboxGL.MapView` (typed as `any`) instead of the direct typed export.
- Installing the main project as a package from the example can be a little finicky, so a workaround is to bundle the parent project with `npm pack`, and then reference that by `npm install`-ing `../<output-archive>`. This behaves exactly as if it were a publicly-available `npm` package.

So to make this ready for `main`, we still need:
1. Fix component type declarations.
2. Type the default `MapboxGL` export.

I can't justify continuing to work on this non-stop, but it's _really_ close, and I'm stuck on (1) above. It would be great if I could get a second pair of eyes. @mfazekas if you're okay with it, I'll merge it into a new `chore/typescript` branch, and then it'll be easier to collaborate on this.

mfazekas commented:
@naftalibeder amazing, thank much! Yes feel free to merge it into a `chore/typescript` branch. 

Doing `npm pack` for example sounds scary for me, we'll need to find a solution for that too.

naftalibeder commented:
> Doing npm pack for example sounds scary for me

Just to be clear, at least for me, the example works perfectly directly referencing the parent project. But I won’t believe it’s a perfectly stable way of doing it until I see a few more people do it successfully :) I mean the `npm pack` option just as a backup strategy.
